### PR TITLE
API: Disallow 0D input arrays in ``nonzero``

### LINKED ISSUE #26238: ENH: Should `np.nonzero` throw exception on 0D input?
### Proposed new feature or change:

The Array API standard says that [`np.nonzero` should throw an exception for 0D input](https://data-apis.org/array-api/latest/API_specification/generated/array_api.nonzero.html).

Currently, `np.nonzero` implicitly makes 0D input a 1D (with `atleast_1d`), ~~no warning is raised~~ a warning is raised. In its docstring it says:
```
       .. deprecated:: 1.17.0

          Use `atleast_1d` explicitly if this behavior is deliberate.
```

Also `np.array(1).nonzero()` gives a deprecation warning:
```
DeprecationWarning: Calling nonzero on 0d arrays is deprecated, as it behaves surprisingly. Use `atleast_1d(arr).nonzero()` if the old behavior was intended.
```

Do we want to throw an exception on 0D input for both cases for Array API compatibility?

### PR DESCRIPTION
**Pull request by mtsokol**
Fixes #26238

This PR disallows 0D arrays input in `np.nonzero` and `np.ndarray.nonzero`.

### CODE REVIEW

File: doc/release/upcoming_changes/26268.improvement.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Scalars and 0D arrays are disallowed for `numpy.nonzero` and `numpy.ndarray.nonzero`.
```
Comment (seberg) [inline doc/release/upcoming_changes/26268.improvement.rst:1]: Should start with a `*` and be named `expired.rst` (IIRC)

**Review by seberg — APPROVED** [review]
Code changes, LGTM.  Would be good to modify the release fragment and rename (we have an expired category).

File: doc/release/upcoming_changes/26268.improvement.rst (hunk 1 above)
Comment (mtsokol) [reply doc/release/upcoming_changes/26268.improvement.rst]: Done!

File: numpy/_core/src/multiarray/item_selection.c [hunk 2]
```diff
@@ -2797,6 +2797,13 @@ NPY_NO_EXPORT PyObject *
 PyArray_Nonzero(PyArrayObject *self)
 {
     int i, ndim = PyArray_NDIM(self);
+    if (ndim == 0) {
+        PyErr_SetString(PyExc_ValueError,
+            "Calling nonzero on 0d arrays is not allowed. "
+            "Use np.atleast_1d(scalar).nonzero() instead.");
```
Comment (ngoldbaum) [inline numpy/_core/src/multiarray/item_selection.c:2803]: Can you leave the slightly nicer error message for bool arrays from the old deprecated warning?

Comment (mtsokol) [reply numpy/_core/src/multiarray/item_selection.c]: Done!

**Comment by ngoldbaum** [comment]
Thanks Mateusz!

**Pull request merged**